### PR TITLE
semgrep 0.49.0

### PR DESCRIPTION
**Pull request by brendongo**
[`action-homebrew-bump-formula`](https://github.com/dawidd6/action-homebrew-bump-formula)

---

Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by carlocab** [comment]
ARM failure happens in this step: https://github.com/Homebrew/homebrew-core/pull/76176/checks?check_run_id=2459974588#step:7:824

Also, no switch to `tree-sitter` dependency yet?

**Comment by brendongo** [comment]
@carlocab we're tied to a specific version of tree-sitter and it doesnt look like the formula has a way of specifying that?

**Comment by brendongo** [comment]
I'm not sure what's happening in the ARM failiure. Looks like some ocaml stdlib things are missing?

**Comment by cho-m** [comment]
Even `--build-from-source` for current **0.48.0** fails on my M1 macOS 11.3.

I wonder if something changed in macOS SDK with 11.3 release.
```console
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:37:
/private/tmp/opamroot20210429-80993-13uj1lr/ocaml-base-compiler.4.10.2/lib/ocaml/version:1:1: error: expected unqualified-id
4.10.2
^
```

**Comment by Bo98** [comment]
Yes, Xcode 12.5 contains the new C++20 `version` header (it previously had a preview version named `__cxx_version`).

**Comment by Bo98** [comment]
Having a file like that be in the include path sounds like a really bad idea. Why is something as broad as `lib/ocaml` with miscellaneous files in the include path?

**Comment by cho-m** [comment]
OCaml Upstream Issue: https://github.com/ocaml/ocaml/issues/9523
Was fixed in **4.12.0**: https://github.com/ocaml/ocaml/commit/b727f422bd48022a830566f19c7d5e50181a96e8

This formula uses **4.10.2** compiler.
The question is whether dependencies allow for using a newer OCaml compiler.
I think the `"lsp" {= "1.1.0"}` is a problem as 1.5.0 is needed for support of 4.12 compiler.
`  - semgrep → lsp < 1.2.0 → ocaml < 4.12`

**Comment by carlocab** [comment]
Probably not, no. We can scope trying to use the 4.12 compiler to ARM, though.

**Comment by brendongo** [comment]
@carlocab worth a try. Is that something I can specify in the formula?

**Comment by carlocab** [comment]
You could try something like this:

```diff
diff --git a/Formula/semgrep.rb b/Formula/semgrep.rb
index e78682d7975..3de5ce3e4a2 100644
--- a/Formula/semgrep.rb
+++ b/Formula/semgrep.rb
@@ -126,8 +126,9 @@ class Semgrep < Formula
       # Used by tree-sitter to place libtree-sitter.a, and header files
       ENV["PREFIX"] = prefix
 
+      compiler_version = (Hardware::CPU.arm?) ? "4.12" : "4.10"
       system "opam", "init", "--no-setup", "--disable-sandboxing"
-      ENV.deparallelize { system "opam", "switch", "create", "ocaml-base-compiler.4.10.2" }
+      ENV.deparallelize { system "opam", "switch", "create", "ocaml-base-compiler.#{compiler_version}" }
 
       system "opam", "exec", "--", "make", "setup"
```

Note, however, that this will stop working on Intel as well as soon as we upgrade our Big Sur Intel nodes to 11.3 (which will be soon).

**Comment by brendongo** [comment]
yep still doesn't work cause of the lsp dependency. We're starting work on updating to ocaml 4.12 but it will probably take a week or two given all the other feature work we have planned. Is there a path forward here or should we just have the homebrew version stay at 0.48.0 til we update to 4.12?

**Comment by Bo98** [comment]
Is it possible to delete the bad file between creating the switch and starting the build?

**Comment by cho-m** [comment]
Deleting OCaml's `version` file should be a possible solution.
I was able to do get formula to build using:
```diff
@@ -128,6 +128,7 @@ class Semgrep < Formula

       system "opam", "init", "--no-setup", "--disable-sandboxing"
       ENV.deparallelize { system "opam", "switch", "create", "ocaml-base-compiler.4.10.2" }
+      rm "#{opamroot}/ocaml-base-compiler.4.10.2/lib/ocaml/version"

       system "opam", "exec", "--", "make", "setup"
```

---

```
🍺  /opt/homebrew/Cellar/semgrep/0.49.0: 1,814 files, 60.9MB, built in 9 minutes 44 seconds
```

**Comment by mjambon** [comment]
It looks like we should upgrade semgrep-core to use lsp.1.5.0 and maybe after that everything will just work with ocaml 4.12. I created an [issue for the lsp upgrade](https://github.com/returntocorp/semgrep/issues/3026). Thank you all for your patience!

**Comment by brendongo** [comment]
> ```diff
> rm "#{opamroot}/ocaml-base-compiler.4.10.2/lib/ocaml/version"
> ```

Going to try as a temp fix while working on 4.12 upgrade

File: Formula/semgrep.rb [hunk 1]
```diff
@@ -127,8 +127,13 @@ def install
...
       system "opam", "init", "--no-setup", "--disable-sandboxing"
-      compiler_version = Hardware::CPU.arm? ? "4.12.0" : "4.10.2"
-      ENV.deparallelize { system "opam", "switch", "create", "ocaml-base-compiler.#{compiler_version}" }
+      ENV.deparallelize { system "opam", "switch", "create", "ocaml-base-compiler.4.10.2" }
+
+      if Hardware::CPU.arm?
+        # Delete OCaml version file since it conflicts with C++20 version header
+        # This can be removed once semgrep upgrades to ocaml 4.12.0
+        rm "#{opamroot}/ocaml-base-compiler.4.10.2/lib/ocaml/version"
+      end
```
Comment (carlocab) [inline Formula/semgrep.rb:136]: Just make the `rm` unconditional. This will be needed on Intel Big Sur too as soon as the runners are upgraded to 11.3.

**Review by carlocab — APPROVED** [review]
🥳

**Pull request closed without merging**